### PR TITLE
Document workarounds

### PR DESCRIPTION
**Pull request by pitr-ch**

### CODE REVIEW

File: doc/contributor/workarounds.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Patching system
+
+The patching system is loaded in `post-boot.rb` and is able to patch stdlib and installed gems. 
+The code implementing the patching itself can be found in `lib/truffle/truffle/patching.rb`.
+It can be disables by `-Xpatching=false` option.
```
Comment (nirvdrum) [inline doc/contributor/workarounds.md]: "can be disables by" -> "can be disabled by passing the"

File: doc/contributor/workarounds.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+Some C extension of gems need to be patched to make it work with TruffleRuby. 
+The patches are defined in `lib/cext/preprocess.rb` 
+and applied immediately before compiling the extension's C source files.
+
+# Legacy workarounds 
+
+They should all be eventually removed, no new usages should be added.
+
+## Truffleruby tool
```
Comment (nirvdrum) [inline doc/contributor/workarounds.md]: "Truffleruby" -> "TruffleRuby"

File: doc/contributor/workarounds.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+Is a legacy tool to setup patches to run gem tests. 
+It's used only in `jt test ecosystem` mainly for Rails testing.
+The tool should be removed and no new tests should use it.
+
+## Constant replacement
+
+`org.truffleruby.parser.ConstantReplacer` replaces constant 
+if a partial file path matches and the name of the constant equals.
+Usually used to fix Ruby implementation detection.
+Should be removed not used further.
```
Comment (nirvdrum) [inline doc/contributor/workarounds.md]: "be removed not" -> "be removed and not"

**Comment by pitr-ch** [comment]
Thanks @nirvdrum, fixed

File: doc/contributor/workarounds.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# Patching system
+
+The patching system is loaded in `post-boot.rb` and is able to patch stdlib and installed gems. 
+The code implementing the patching itself can be found in `lib/truffle/truffle/patching.rb`.
+It can be disabled by passing the `-Xpatching=false` option.
+
+The patching system works for gems as follows. 
+When a gem *g* is activated and there is a directory `g` in `lib/patches`, 
+the directory is injected into `LOAD_PATH` before the original load-paths of the *g* gem.
```
Comment (eregon) [inline doc/contributor/workarounds.md]: `$LOAD_PATH`

File: doc/contributor/workarounds.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+The patching system is loaded in `post-boot.rb` and is able to patch stdlib and installed gems. 
+The code implementing the patching itself can be found in `lib/truffle/truffle/patching.rb`.
+It can be disabled by passing the `-Xpatching=false` option.
+
+The patching system works for gems as follows. 
+When a gem *g* is activated and there is a directory `g` in `lib/patches`, 
+the directory is injected into `LOAD_PATH` before the original load-paths of the *g* gem. 
+As a result the patching files are loaded first before the original files in the gem.
+The patching file is responsible for loading the original file (if desirable),
+which can be easily done with helper method `Truffle::Patching.require_original __FILE__`.
```
Comment (eregon) [inline doc/contributor/workarounds.md]: > which can be done with `Truffle::Patching.require_original __FILE__`

seems simpler

File: doc/contributor/workarounds.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
+# Patching system
+
+The patching system is loaded in `post-boot.rb` and is able to patch stdlib and installed gems. 
+The code implementing the patching itself can be found in `lib/truffle/truffle/patching.rb`.
+It can be disabled by passing the `-Xpatching=false` option.
+
+The patching system works for gems as follows. 
+When a gem *g* is activated and there is a directory `g` in `lib/patches`,
```
Comment (eregon) [inline doc/contributor/workarounds.md]: Additionally the directory needs to be listed in `patching.rb`.

File: doc/contributor/workarounds.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+The patching system works for gems as follows. 
+When a gem *g* is activated and there is a directory `g` in `lib/patches`, 
+the directory is injected into `LOAD_PATH` before the original load-paths of the *g* gem. 
+As a result the patching files are loaded first before the original files in the gem.
+The patching file is responsible for loading the original file (if desirable),
+which can be easily done with helper method `Truffle::Patching.require_original __FILE__`.
+
+# C file preprocessing
+
+Some C extension of gems need to be patched to make it work with TruffleRuby.
```
Comment (eregon) [inline doc/contributor/workarounds.md]: to make them work

File: doc/contributor/workarounds.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+the directory is injected into `LOAD_PATH` before the original load-paths of the *g* gem. 
+As a result the patching files are loaded first before the original files in the gem.
+The patching file is responsible for loading the original file (if desirable),
+which can be easily done with helper method `Truffle::Patching.require_original __FILE__`.
+
+# C file preprocessing
+
+Some C extension of gems need to be patched to make it work with TruffleRuby. 
+The patches are defined in `lib/cext/preprocess.rb` 
+and applied immediately before compiling the extension's C source files.
```
Comment (eregon) [inline doc/contributor/workarounds.md]: and applied immediately on the C source before passing it to the compiler.

File: doc/contributor/workarounds.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+# Legacy workarounds 
+
+They should all be eventually removed, no new usages should be added.
```
Comment (eregon) [inline doc/contributor/workarounds.md]: `,` => `and` maybe?

File: doc/contributor/workarounds.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+## TruffleRuby tool
+
+Is a legacy tool to setup patches to run gem tests. 
+It's used only in `jt test ecosystem` mainly for Rails testing.
+The tool should be removed and no new tests should use it.
```
Comment (eregon) [inline doc/contributor/workarounds.md]: Redundant with the first paragraph above.

File: doc/contributor/workarounds.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+## Constant replacement
+
+`org.truffleruby.parser.ConstantReplacer` replaces constant 
+if a partial file path matches and the name of the constant equals.
+Usually used to fix Ruby implementation detection.
+Should be removed and not used further.
```
Comment (eregon) [inline doc/contributor/workarounds.md]: Redundant with the first paragraph above.

File: doc/contributor/workarounds.md (hunk 10 above)
Comment (chrisseaton) [reply doc/contributor/workarounds.md]: I think it's fine to restate.

**Comment by chrisseaton** [comment]
Thanks this clears up a lot for me. Benoit's English fixes are useful though.

File: doc/contributor/workarounds.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+The patching system is loaded in `post-boot.rb` and is able to patch stdlib and installed gems. 
+The code implementing the patching itself can be found in `lib/truffle/truffle/patching.rb`.
+It can be disabled by passing the `-Xpatching=false` option.
+
+The patching system works for gems as follows. 
+When a gem *g* is activated, 
+there is a directory `g` in `lib/patches`, 
+and the directory is listed in `lib/truffle/truffle/patching.rb`,  
+the directory is injected into `$LOAD_PATH` before the original load-paths of the *g* gem.
```
Comment (eregon) [inline doc/contributor/workarounds.md]: Minor, `inserted` here might be better than `injected`, implicitly referring to Array#insert.

**Pull request merged**